File: Week_02/589/n_ary_tree_preorder_traversal.py
from typing import List


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []

        def pre_order(node: Node):
            if node is not None:
                res.append(node.val)
                for c in node.children:
                    pre_order(c)

        pre_order(root)
        return res

File: Week_02/589/n_ary_tree_preorder_traversal.go
package main

type Node struct {
	Val      int
	Children []*Node
}

func preorder(root *Node) []int {
	res := make([]int, 0)
	var preOrder func(node *Node)
	preOrder = func(node *Node) {
		if node != nil {
			res = append(res, node.Val)
			for i := 0; i < len(node.Children); i++ {
				preOrder(node.Children[i])
			}
		}
	}
	preOrder(root)
	return res
}
